### ISSUE TITLE
Allow datagram send order relative to streams to be application defined

### ISSUE THREAD

**Issue by wilaw** [w3c/webtransport#451]
PR #438 defined an optional send order for streams. This allows relative prioritization between streams, however it does not define the relative prioritization between datagrams and streams. 

Consider the use case of a streaming protocol in which a bidi stream is established as a control channel and A/V data is transmitted via datagrams. The current UA implementation always gives datagrams a higher send priority over all streams. In the face of congestion, the A/V data would be sent ahead of any control data.

One solution to this problem would be to add an optional sendOrder attribute to the datagrams.writable.getWriter method:

```
const wt = new WebTransport(url);
const writer = wt.datagrams.writable.getWriter({sendOrder:3});
```

Another might be to add a datagramSendOrder entry to the WebTransportOptions dictionary

```
const wt = new WebTransport(url,{datagramSendOrder:3});
const writer = wt.datagrams.writable.getWriter();
```

If datagram sendOrder is left undefined, then the default prioritization can be either that:
1. Datagrams are always sent ahead of any queued stream data
2. Datagrams are round-robined with other queued stream data that also has an undefined sendOrder.

**Comment by bdrtc** [comment]
do we need support packet/frame level priority ?
one use case is retransmitte packet may have high prioriy than nomal packet.

**Comment by wilaw** [comment]
@bdrtc - packet/frame prioritization is controlled at the WebTransport protocol level, which is managed by the browser. The send order referenced in this issue refers to the order in which the send queues are drained by the browser and hence exists at a level above the protocol layer which you are referencing here. Suggest that you open a new issue that indicates why a JS application would need to control the lower level protocol characteristics. In the past, the browser implementers have been reticent in allowing application interference in protocol stack management, but you may ask again anyway.

**Comment by bdrtc** [comment]
> 1. Datagram

hi @wilaw , 
packet level prioritization here is also the order in which the send queues are drained by the browser.
one  rtc use case:

when use webtransport datagram transfer audio/video rtp packet in rtc ,  users need impliment retransmitte/redundancy（fec） logic in js level as webtransport datagram is unreliable,  but in rtc case,  audio/video data need partially reliable, retransmitte/fec packet here is js in level and user defined. they need different  prioritization with normal audio/video packet 
in js level.

**Comment by wilaw** [comment]
Since this issue was raised, we have merged the concept of sendGroups. How do these groups relate to datagrams? This seems to be implementation defined behavior currently. 
1. Are datagrams in an implicit SendGroup?
2. Do all datagrams have the same implicit SendOrder within that SendGroup?

If [1] is true, then all datagram traffic will be fair-shared with any other stream traffic. Is this acceptable, or do we need weighting between SendGroups?

**Comment by jan-ivar** [comment]
> 1. Are datagrams in an implicit SendGroup?

I agree this is implementation-defined today. It used to say datagrams got priority, but this was removed in https://github.com/w3c/webtransport/pull/394.

> 2. Do all datagrams have the same implicit SendOrder within that SendGroup?

SendOrder was created to disambiguate order between send-streams, which have a creation order, but exist in parallel.

Datagrams OTOH have an implicit order from being written to a writable, and are sent one after the other AFAIK.

SendOrder also allowed application tricks like jumping the queue, though it remains unclear to me whether that behavior alone would have been enough to specify it, or whether we desire that behavior for datagrams. Datagrams already support [outgoingMaxAge](https://w3c.github.io/webtransport/#dom-webtransportdatagramduplexstream-outgoingmaxage) which is probably a better solution for managing urgency.

**Comment by jan-ivar** [comment]
Meeting:
- More discussion needed.

**Comment by jan-ivar** [comment]
Meeting:
- Appears that Firefox (datagrams lowest priority) and Chrome (datagrams have highest priority) differ here, which is unfortunate 
- options for datagrams seem to be: first priority, last priority, round-robin
- testable? all probably are, so picking based on testing may not be necessary.
- Would be desirable if implementations followed similar implementation approach to how they prioritize sending
- Continue discussion when @ekinnear is present

**Comment by wilaw** [comment]
How about we add this language

> The sending of datagrams SHOULD be given higher priority than send-streams but MUST NOT starve them.

**Comment by jan-ivar** [comment]
Meeting:
- If datagrams have a lower priority, that encourages devs to avoid using streams to ensure their datagrams do arrive
- If datagrams have a higher priority, dev baseline expectations of UDP are applicable, you can use all the streams you want without having it restrict your datagrams.
- 100% starving seems undesirable, we need streams to make some progress
- conclusion: @wilaw's proposal above
